### ISSUE TITLE
nrows support in cudf.io.csv.read_csv_strings() is causing an exception

### ISSUE THREAD

**Issue by kenhester** [rapidsai/cudf#1049]
**Describe the bug**
When calling cudf.io.csv.read_csv_strings() using nrows to read a file in chunks, I get a thrust exception.

Exception:
```
nvs-idx: computing sizes: cudaErrorIllegalAddress(77):an illegal memory access was encountered
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  device free failed: an illegal memory access was encountered
```

**Steps/Code to reproduce bug**
Full python code in file attached

Snippet:
```
while True:
    frame = cudf.io.csv.read_csv_strings(file, quoting=False, doublequote=False, names=names, dtype=dtypes, skiprows=skip_rows, nrows=rows_to_read)
    skip_rows += rows_to_read
...
    if rows_read < rows_to_read:
        break

```

**Expected behavior**
Read the file in chunk windows.  For last chunk, return remaining rows without exception.

**Environment details (please complete the following information):**
Latest NGC RAPIDS CUDA 10 Container (0.5.1)
 

**Additional context**
Data file is large and private; provided on request.

Program File:
[load_pos_file.zip](https://github.com/rapidsai/cudf/files/2906448/load_pos_file.zip)

**Comment by j-ieong** [comment]
@kenhester, can you try 0.6?
I think some parameters were not handled the same way between `read_csv()` and `read_csv_strings()` that was fixed in 0.6.

**Comment by kenhester** [comment]
If I provide a data file, would you be able to test the fix applied in 0.6?

Ken Hester | NVIDIA Solutions Architect
c: 832-216-8736
khester@nvidia.com<mailto:khester@nvidia.com>

GTC 2019<https://www.nvidia.com/en-us/gtc/> |San Jose Convention Center|Workshops- March 17, 2019  |Conference- March 18-21,2019

**Comment by vuule** [comment]
Yes. If you provide the code to repro the issue and the data file, we can check it against 0.6.

**Comment by kenhester** [comment]
Code is attached to the issue in the zip file.

I can provide access to a data file through a private channel as the data not public.

Ken Hester | NVIDIA Solutions Architect
c: 832-216-8736
khester@nvidia.com<mailto:khester@nvidia.com>

GTC 2019<https://www.nvidia.com/en-us/gtc/> |San Jose Convention Center|Workshops- March 17, 2019  |Conference- March 18-21,2019

**Comment by vuule** [comment]
Getting repro even with a small byte_range. It's starting to look like the issue is not related to OOM. 
I'll isolate further today and update.

**Comment by vuule** [comment]
Amazingly, I think I managed to reduce the repro to the following:
    repro_buffer =       'abcdefghijkl\n'+'abcdefghijk\n'
    no_repro_buffer = 'abcdefghijkl\n'+'abcdefghij\n'
    frame = read_csv(StringIO(*_buffer), header=None)
In the repro case, there are 25 characters in total. With 24 or fewer characters, there is no repro. The crash is not 100% consistent, but it is still very frequent (>50%).

**Comment by vuule** [comment]
I have locally resolved the issue in the previous comment. Unfortunately, it is not related to the issue reported here.

I'm getting a thrust error when using nrows to parse ~1GB of data. The error only happens when the data types are explicitly set. With data type inference, everything works fine. This isolates the issue to nvstring fields. Looking into the root cause now.

**Comment by vuule** [comment]
@kenhester can you please try using byte_range with a ~1GB range, and quotechar='\0' (or whatever option to turn off quotation)?
On my system it's hard to tell if it runs out of memory or not, but it should work for you.
I'll check tomorrow if there is any error checking missing.

**Comment by vuule** [comment]
While looking for a minimal repro, I found that this file has unmatched quotes, probably because of the use of unicode character for the open quote. Because of this, we cannot expect read_csv to work correctly with this input, when quoting is enabled.

With quoting disabled, and with the use of byte_range to limit the memory usage, I'm not getting any errors with this file. 
@mjsamoht I think this issue should be closed, unless Ken has a repro with applicable parameters.

@kenhester ping for info about the repro with quoting disabled.

**Comment by mjsamoht** [comment]
I moved this issue to P2 for time being.  It doesn't look like a blocker for v0.6, and perhaps not a bug at all since it uses unicode and unmatched quotations.

**Comment by kenhester** [comment]
Confirming: disable quoting allows the file to read to the end without exception.

**Comment by mjsamoht** [comment]
We may still want to ensure there is no exception and have read_csv return more gracefully.  But this won't make 0.6.

**Comment by randerzander** [comment]
The title of this issue is about `read_csv_strings` which will be removed in 0.6.

However, the comments suggest there may be similar parsing problem with `read_csv`.

@mjsamoht @kenhester can we either close this bug or update the title/description to reflect the issue with `read_csv`?

**Comment by vuule** [comment]
Tried wrapping the code that throws the error in a catch-all block. The terminate() call still happens.

Looking at the implementation of NVStrings::replace(), I see that on CUDA error, the function continues execution after printing the error and proceeds to deallocate CUDA buffers. The terminate() call originates in one of these calls. Because of this, I cannot change the behavior from the read_csv function, this is an issue in NVStrings.
@mjsamoht Can we move this issue to custrings project?

**Comment by mjsamoht** [comment]
@kkraus14 how do we track NVStrings related issues?  Is there a special label?

**Comment by mjsamoht** [comment]
This issue needs to be moved to https://github.com/rapidsai/custrings.  I'm not sure how to do that though.
